USF-1670: Dropins Localization (#95)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/all/labeling.mdx

import CodeInclude from '@components/CodeInclude.astro';
import dictionary from '@dropins/tools/types/elsie/src/i18n/en_US.json.d.ts?raw';
import Vocabulary from '@components/Vocabulary.astro';
import Callouts from '@components/Callouts.astro';
import OptionsTable from '@components/OptionsTable.astro';
import { Steps } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';

In this topic, you will learn how to update drop-in component labels by using the resources available in the commerce boilerplate.

## Key concepts

<Vocabulary>

### Placeholders file

A Google Sheets file (`placeholders`) or a SharePoint Excel file (`placeholders.xlsx`) where you change and maintain your storefront's labels, which includes localization. See the [Placeholders file](/resources/placeholders/) for more information.

### Language objects

A JavaScript object (named `langDefinitions` by convention) that contains key-value pairs for UI text labels in a specific language. The object is used to override the default dictionary with the placeholder file's keys and values.

### Drop-in component dictionaries

> JSON files that define the shape, keys, and default values for labels and text within a drop-in component. Each drop-in component has a default dictionary file: `i18n/en_US.json`. But drop-in components can contain multiple dictionary files as the site is localized for other languages, including `fr_CA.json` for French, `de_DE.json` for German, and so on. The documentation for each drop-in component includes the default dictionary for that drop-in component. As an example, see the Product details page [Dictionary](/dropins/product-details/dictionary) topic.

Reviewer: ```suggestion
JSON files that define the shape, keys, and default values for labels and text within a drop-in component. Each drop-in component has a default dictionary file: `i18n/en_US.json`. But drop-in components can contain multiple dictionary files as the site is localized for other languages, including `fr_CA.json` for French, `de_DE.json` for German, and so on. The documentation for each drop-in component includes the default dictionary for that drop-in component. As an example, see the [dictionary](/dropins/product-details/dictionary) for the product details page drop-in component.
```

1. **Import function**. You need to import the `fetchPlaceholders` function from the boilerplate's `aem.js` file.
1. **Fetch placeholders.** Use the `fetchPlaceholders` function to retrieve the `placeholders` key-value pairs from Google Docs or SharePoint folders.
1. **Override default dictionary**. Override the `default` property from the `langDefinitions` object with the keys and values from the `placeholder` object.
1. **Initialize dictionary**. Use the `register` function to update the dictionary at runtime.

</Callouts>

## Step-by-step

In the boilerplate code, the UI text labels in drop-in components come from the `placeholders` file. Using this file as the source for all storefront UI labels across drop-in components makes it easy for merchants to make label changes without developer involvement.

<Tasks>

<Task>
### Import `fetchPlaceholders` function

In the drop-in block (for example, `product-details.js`, `cart.js`), import the `fetchPlaceholders` function from the boilerplate's `aem.js` file.

```javascript
import { fetchPlaceholders } from '../../scripts/aem.js';
```
</Task>

<Task>
### Fetch placeholders

> Within the `decorate` function, Use the `fetchPlaceholders` function to retrieve the `placeholders` language object from Google Docs or SharePoint folders. The following code snippet shows how to fetch the placeholders.

Reviewer: ```suggestion
Within the `decorate` function, use the `fetchPlaceholders` function to retrieve the `placeholders` language object from the Google Docs or SharePoint folders. The following code snippet shows how to fetch the placeholders.
```

```javascript
// Retrieve the placeholders language object
const placeholders = await fetchPlaceholders();
```
</Task>

<Task>
### Override default dictionary

> Within the `decorate` function, override the drop-in component's default dictionary with the placeholders object. The `JSON` code for each drop-in dictionary can be found in Dictionary topic.

Reviewer: ```suggestion
Within the `decorate` function, override the drop-in component's default dictionary with the `placeholders` object. The documentation for each drop-in component includes a "dictionary" topic where you can find `JSON` examples.
```

Reviewer: You might not need the second sentence.

Author: Deleted it.
